AppTaskQueue/appscale/taskqueue/templates/task.py

```diff
@@ -90,42 +96,51 @@ def get_wait_time(retries, args):
     for header in headers:
       connection.putheader(header, headers[header])
 
-    content_length = "0"
-    if args["body"]:
-      content_length = str(len(args['body']))
-
     if 'content-type' not in headers or 'Content-Type' not in headers:
       if url.query:
         connection.putheader('content-type', 'application/octet-stream')
       else:
         connection.putheader('content-type',
                              'application/x-www-form-urlencoded')
 
-    connection.putheader("Content-Length", content_length)
-    connection.endheaders()
-    if args["body"]:
-      connection.send(args['body'])
-    response = connection.getresponse()
-    payload = response.read()
-    response.close()
+    connection.putheader("Content-Length", str(content_length))
+
     retries = int(QUEUE_NAME.request.retries) + 1
+    wait_time = get_wait_time(retries, args)
+
+    try:
+      connection.endheaders()
+      if args["body"]:
+        connection.send(args['body'])
+
+      response = connection.getresponse()
+      response.read()
+      response.close()
+    except (BadStatusLine, SocketError):
+      logger.warning(
+        '{task} failed before receiving response. It will retry in {wait} '
+        'seconds.'.format(task=args['task_name'], wait=wait_time))
+      raise QUEUE_NAME.retry(countdown=wait_time)
+
     if 200 <= response.status < 300:
       # Task successful.
       item = TaskName.get_by_key_name(args['task_name'])
       db.delete(item)
+      logger.info(
+        '{task} received status {status} from {url}'.format(
+          task=args['task_name'], status=response.status, url=url))
       return response.status
     elif response.status == 302:
       redirect_url = response.getheader('Location')
       logger.info("Task %s asked us to redirect to %s, so retrying there." % (
         args['task_name'], redirect_url))
       url = urlparse(redirect_url)
       if redirects_left == 0:
-        raise QUEUE_NAME.retry(countdown=get_wait_time(retries, args))
+        raise QUEUE_NAME.retry(countdown=wait_time)
       redirects_left -= 1
     else:
       # Fail
       # TODO: Update the database with the failed status
```

Comment:
If I understand correctly, we don't keep track of failed status on purpose (since we do a retry). Is this TODO here for a reason then?